File: bindings/c/test_ex1.c
/*
 * Test Ex1: alpha-208Pb Optical Potential (C)
 * ============================================
 *
 * This replicates the Fortran Ex1 example exactly.
 * Compares results with the Fortran reference.
 *
 * Reference: Goldring et al., Phys. Lett. B32 (1970) 465
 *
 * Compile:
 *   gcc -o test_ex1 test_ex1.c -L../../lib -lhprmat -lm
 *
 * Run:
 *   DYLD_LIBRARY_PATH=../../lib ./test_ex1   (macOS)
 *   LD_LIBRARY_PATH=../../lib ./test_ex1     (Linux)
 */

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <complex.h>
#include "hprmat.h"

int main() {
    printf("============================================================\n");
    printf("HPRMAT C Test: Ex1 alpha-208Pb Optical Potential\n");
    printf("============================================================\n");

    /* Physical constants (same as Fortran) */
    double a1 = 208.0, a2 = 4.0;    /* Mass numbers */
    double z1 = 82.0, z2 = 2.0;     /* Charge numbers */
    double rmu = a1 * a2 / (a1 + a2);  /* Reduced mass */
    double ze = z1 * z2 * 1.44;     /* Z1*Z2*e^2 (MeV.fm) */
    double hm = 20.736 / rmu;       /* hbar^2/(2*mu) in MeV.fm^2 */
    double eta0 = ze / (2 * hm);    /* Sommerfeld parameter factor */

    /* R-matrix parameters (from data1) */
    int l = 20;        /* Angular momentum */
    int nr = 60;       /* Lagrange functions per interval */
    int ns = 1;        /* Number of intervals */
    double rmax = 14.0;/* Channel radius (fm) */

    /* Energy parameters */
    int ne = 5;        /* Number of energies */
    double e0 = 10.0;  /* Initial energy (MeV) */
    double estep = 10.0; /* Energy step (MeV) */

    /* Woods-Saxon parameters */
    double an = 0.5803;  /* Diffuseness (fm) */
    double rn = 1.1132 * (pow(a1, 1.0/3) + pow(a2, 1.0/3));  /* Radius (fm) */

    printf("\nParameters:\n");
    printf("  Angular momentum L = %d\n", l);
    printf("  Lagrange functions: nr = %d, ns = %d\n", nr, ns);
    printf("  Channel radius: rmax = %.1f fm\n", rmax);
    printf("  Reduced mass: mu = %.4f amu\n", rmu);
    printf("  hbar^2/(2*mu) = %.6f MeV.fm^2\n", hm);

    /* Initialize R-matrix */
    int ntot = nr * ns;
    double *zrma = (double*)malloc(ntot * sizeof(double));
    hprmat_init(nr, ns, rmax, zrma);

    printf("  Mesh points: %d\n", ntot);
    printf("  First mesh point: r = %.6f fm\n", zrma[0]);
    printf("  Last mesh point: r = %.6f fm\n", zrma[ntot-1]);

    /* Build potential matrix (complex Woods-Saxon + Coulomb) */
    int nch = 1;
    double _Complex *cpot = (double _Complex*)calloc(ntot * nch * nch, sizeof(double _Complex));

    for (int i = 0; i < ntot; i++) {
        double r = zrma[i];
        double xx = 1.0 + exp((r - rn) / an);
        double _Complex cvn = (-100.0 - 10.0*I) / xx;  /* Complex Woods-Saxon */
        double vc = ze / r;                             /* Coulomb */
        cpot[i] = (cvn + vc) / hm;  /* cpot[i, 0, 0] in column-major */
    }

    /* Channel parameters */
    int lval[1] = {l};

    printf("\nEnergies: %d points from %.1f to %.1f MeV\n", ne, e0, e0 + (ne-1)*estep);
    printf("------------------------------------------------------------\n");
    printf("%10s %14s %14s %10s\n", "E (MeV)", "Re(S)", "Im(S)", "|S|");
    printf("------------------------------------------------------------\n");

    /* Output arrays */
    double _Complex *cu = (double _Complex*)calloc(nch * nch, sizeof(double _Complex));
    int nopen;

    /* Test all solver types */
    const char *solver_names[] = {"", "Dense LAPACK", "Mixed Precision", "Woodbury"};

    for (int solver = 1; solver <= 3; solver++) {
        printf("\n--- Solver %d: %s ---\n", solver, solver_names[solver]);

        for (int ie = 0; ie < ne; ie++) {
            double ecm = e0 + ie * estep;

            /* Wave number and Sommerfeld parameter */
            double qk_val = sqrt(ecm / hm);
            double eta_val = eta0 / qk_val;

            double qk[1] = {qk_val};
            double eta[1] = {eta_val};

            /* Solve */
            hprmat_solve(nch, lval, qk, eta, rmax, nr, ns,
                         cpot, ntot, cu, &nopen, solver);

            /* Extract S-matrix element */
            double _Complex S11 = cu[0];
            printf("%10.3f %14.6e %14.6e %10.6f\n",
                   ecm, creal(S11), cimag(S11), cabs(S11));
        }
    }

    printf("------------------------------------------------------------\n");
    printf("\nTest completed!\n");

    /* Clean up */
    free(zrma);
    free(cpot);
    free(cu);

    return 0;
}

File: bindings/python/test_ex1.py
#!/usr/bin/env python3
"""
Test Ex1: alpha-208Pb Optical Potential
========================================

This replicates the Fortran Ex1 example exactly.
Compares results with the Fortran reference.

Reference: Goldring et al., Phys. Lett. B32 (1970) 465
"""

import numpy as np
import sys

try:
    import hprmat_fortran as hpf
except ImportError:
    print("Error: hprmat_fortran module not found.")
    print("Build it first with f2py")
    sys.exit(1)


def main():
    print("=" * 60)
    print("HPRMAT Python Test: Ex1 alpha-208Pb Optical Potential")
    print("=" * 60)

    # Physical constants (same as Fortran)
    a1, a2 = 208.0, 4.0    # Mass numbers
    z1, z2 = 82.0, 2.0     # Charge numbers
    rmu = a1 * a2 / (a1 + a2)  # Reduced mass
    ze = z1 * z2 * 1.44    # Z1*Z2*e^2 (MeV.fm)
    hm = 20.736 / rmu      # hbar^2/(2*mu) in MeV.fm^2
    eta0 = ze / (2 * hm)   # Sommerfeld parameter factor

    # R-matrix parameters (from data1)
    l = 20       # Angular momentum
    nr = 60      # Lagrange functions per interval
    ns = 1       # Number of intervals
    rmax = 14.0  # Channel radius (fm)

    # Energy parameters
    ne = 5       # Number of energies
    e0 = 10.0    # Initial energy (MeV)
    estep = 10.0 # Energy step (MeV)

    # Woods-Saxon parameters
    an = 0.5803  # Diffuseness (fm)
    rn = 1.1132 * (a1**(1.0/3) + a2**(1.0/3))  # Radius (fm)

    print(f"\nParameters:")
    print(f"  Angular momentum L = {l}")
    print(f"  Lagrange functions: nr = {nr}, ns = {ns}")
    print(f"  Channel radius: rmax = {rmax} fm")
    print(f"  Reduced mass: mu = {rmu:.4f} amu")
    print(f"  hbar^2/(2*mu) = {hm:.6f} MeV.fm^2")

    # Initialize R-matrix
    zrma = hpf.py_rmat_init(nr, ns, rmax)
    ntot = nr * ns

    print(f"  Mesh points: {ntot}")
    print(f"  First mesh point: r = {zrma[0]:.6f} fm")
    print(f"  Last mesh point: r = {zrma[-1]:.6f} fm")

    # Build potential matrix (complex Woods-Saxon + Coulomb)
    nch = 1
    cpot = np.zeros((ntot, nch, nch), dtype=np.complex128, order='F')

    for i in range(ntot):
        r = zrma[i]
        xx = 1.0 + np.exp((r - rn) / an)
        cvn = complex(-100.0, -10.0) / xx  # Complex Woods-Saxon
        vc = ze / r                         # Coulomb
        cpot[i, 0, 0] = (cvn + vc) / hm     # Divided by hbar^2/(2*mu)

    # Channel parameters
    lval = np.array([l], dtype=np.int32)

    print(f"\nEnergies: {ne} points from {e0} to {e0 + (ne-1)*estep} MeV")
    print("-" * 60)
    print(f"{'E (MeV)':>10} {'Re(S)':>14} {'Im(S)':>14} {'|S|':>10}")
    print("-" * 60)

    # Test all solver types
    for solver in [1, 2, 3]:
        solver_names = {1: "Dense LAPACK", 2: "Mixed Precision", 3: "Woodbury"}
        print(f"\n--- Solver {solver}: {solver_names[solver]} ---")

        for ie in range(ne):
            ecm = e0 + ie * estep

            # Wave number and Sommerfeld parameter
            qk_val = np.sqrt(ecm / hm)
            eta_val = eta0 / qk_val

            qk = np.array([qk_val], dtype=np.float64)
            eta = np.array([eta_val], dtype=np.float64)

            # Solve
            cu, nopen = hpf.py_rmatrix(nch, ntot, lval, qk, eta, rmax, nr, ns, cpot, solver)

            # Extract S-matrix element
            S11 = cu[0, 0]
            print(f"{ecm:10.3f} {S11.real:14.6e} {S11.imag:14.6e} {abs(S11):10.6f}")

    print("-" * 60)
    print("\nTest completed!")


if __name__ == "__main__":
    main()
